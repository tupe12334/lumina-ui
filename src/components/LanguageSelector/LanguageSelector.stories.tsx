import type { Meta, StoryObj } from '@storybook/react'
import { LanguageSelector } from './LanguageSelector'

const meta: Meta<typeof LanguageSelector> = {
  title: 'Components/LanguageSelector',
  component: LanguageSelector,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes: {
    size: {
      control: 'select',
      options: ['sm', 'md', 'lg'],
    },
    variant: {
      control: 'select',
      options: ['default', 'error', 'success'],
    },
    disabled: {
      control: 'boolean',
    },
  },
}

export default meta
type Story = StoryObj<typeof meta>

export const Default: Story = {
  args: {
    placeholder: 'Select a language',
  },
}

export const WithSelectedLanguage: Story = {
  args: {
    value: 'en',
    placeholder: 'Select a language',
  },
}

export const WithoutFlags: Story = {
  args: {
    placeholder: 'Select a language',
    config: {
      showFlags: false,
    },
  },
}

export const WithRegions: Story = {
  args: {
    placeholder: 'Select a language',
    config: {
      showFlags: true,
      showRegions: true,
    },
  },
}

export const WithRegionInLabel: Story = {
  args: {
    placeholder: 'Select a language',
    config: {
      showFlags: true,
      includeRegionInLabel: true,
    },
  },
}

export const Searchable: Story = {
  args: {
    placeholder: 'Search for a language',
    config: {
      searchable: true,
      showFlags: true,
    },
  },
}

export const Clearable: Story = {
  args: {
    value: 'fr',
    placeholder: 'Select a language',
    config: {
      clearable: true,
      showFlags: true,
    },
  },
}

export const CustomLanguages: Story = {
  args: {
    placeholder: 'Select a programming language',
    config: {
      showFlags: false,
      customLanguages: [
        { code: 'js', name: 'JavaScript', nativeName: 'JavaScript' },
        { code: 'ts', name: 'TypeScript', nativeName: 'TypeScript' },
        { code: 'py', name: 'Python', nativeName: 'Python' },
        { code: 'java', name: 'Java', nativeName: 'Java' },
        { code: 'cpp', name: 'C++', nativeName: 'C++' },
        { code: 'rust', name: 'Rust', nativeName: 'Rust' },
        { code: 'go', name: 'Go', nativeName: 'Go' },
      ],
    },
  },
}

export const Small: Story = {
  args: {
    size: 'sm',
    placeholder: 'Select a language',
  },
}

export const Large: Story = {
  args: {
    size: 'lg',
    placeholder: 'Select a language',
  },
}

export const ErrorState: Story = {
  args: {
    variant: 'error',
    placeholder: 'Select a language',
  },
}

export const Disabled: Story = {
  args: {
    disabled: true,
    placeholder: 'Select a language',
  },
}