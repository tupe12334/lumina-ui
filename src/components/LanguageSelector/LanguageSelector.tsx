import React from 'react'
import { Selector, SelectorOption } from '../Selector/Selector'

export interface Language {
  code: string
  name: string
  nativeName: string
  flag?: string
}

export interface SimpleLanguageSelectorProps {
  languages: Language[]
  value?: string
  placeholder?: string
  disabled?: boolean
  className?: string
  showFlags?: boolean
  onChange?: (languageCode: string) => void
}


export const LanguageSelector = React.forwardRef<HTMLDivElement, SimpleLanguageSelectorProps>(
  (props, ref) => {
    const {
      languages,
      value,
      placeholder,
      disabled,
      className,
      showFlags,
      onChange
    } = props

    const finalLanguages = languages
    const shouldShowFlags = showFlags !== false

    const options: SelectorOption[] = finalLanguages.map((language) => {
      let label = `${language.name}`

      if (language.nativeName !== language.name) {
        label += ` (${language.nativeName})`
      }

      if (shouldShowFlags && language.flag) {
        label = `${language.flag} ${label}`
      }

      return {
        value: language.code,
        label
      }
    })

    return (
      <Selector
        ref={ref}
        options={options}
        value={value}
        placeholder={placeholder || 'Select a language'}
        disabled={disabled}
        className={className}
        onChange={onChange}
      />
    )
  }
)

LanguageSelector.displayName = 'LanguageSelector'