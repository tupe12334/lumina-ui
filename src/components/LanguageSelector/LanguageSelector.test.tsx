import { render, screen, fireEvent } from '@testing-library/react'
import { test, expect, vi } from 'vitest'
import { LanguageSelector } from './LanguageSelector'

const mockLanguages = [
  { code: 'en', name: 'English', nativeName: 'English' },
  { code: 'es', name: 'Spanish', nativeName: 'Español' },
  { code: 'fr', name: 'French', nativeName: 'Français' },
]

test('LanguageSelector renders correctly with default props', () => {
  const { container } = render(<LanguageSelector languages={mockLanguages} />)
  expect(container.firstChild).toMatchSnapshot()
})

test('LanguageSelector renders correctly with selected language', () => {
  const { container } = render(<LanguageSelector languages={mockLanguages} value="en" />)
  expect(container.firstChild).toMatchSnapshot()
})

test('LanguageSelector renders correctly without flags', () => {
  const { container } = render(
    <LanguageSelector
      languages={mockLanguages}
      showFlags={false}
    />
  )
  expect(container.firstChild).toMatchSnapshot()
})


test('LanguageSelector renders correctly when disabled', () => {
  const { container } = render(<LanguageSelector languages={mockLanguages} disabled />)
  expect(container.firstChild).toMatchSnapshot()
})

test('LanguageSelector calls onChange with language code', () => {
  const handleChange = vi.fn()
  render(<LanguageSelector languages={mockLanguages} onChange={handleChange} />)

  const trigger = screen.getByRole('button')
  fireEvent.click(trigger)

  const option = screen.getByText(/English/)
  fireEvent.click(option)

  expect(handleChange).toHaveBeenCalledWith('en')
})


test('LanguageSelector renders correctly with custom languages', () => {
  const customLanguages = [
    { code: 'js', name: 'JavaScript', nativeName: 'JavaScript' },
    { code: 'py', name: 'Python', nativeName: 'Python' },
  ]

  const { container } = render(
    <LanguageSelector
      languages={customLanguages}
      showFlags={false}
    />
  )
  expect(container.firstChild).toMatchSnapshot()
})


test('LanguageSelector renders correctly with custom className', () => {
  const { container } = render(
    <LanguageSelector languages={mockLanguages} className="custom-class" />
  )
  expect(container.firstChild).toMatchSnapshot()
})