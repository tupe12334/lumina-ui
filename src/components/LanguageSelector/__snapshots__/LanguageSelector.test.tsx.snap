// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LanguageSelector renders correctly when disabled 1`] = `
<div
  class="_selector_729a27"
>
  <button
    class="_trigger_729a27"
    disabled=""
    type="button"
  >
    <span
      class="_value_729a27"
    >
      Select a language
    </span>
    <span
      aria-hidden="true"
      class="_icon_729a27"
    >
      â–¼
    </span>
  </button>
</div>
`;

exports[`LanguageSelector renders correctly with custom className 1`] = `
<div
  class="_selector_729a27 custom-class"
>
  <button
    class="_trigger_729a27"
    type="button"
  >
    <span
      class="_value_729a27"
    >
      Select a language
    </span>
    <span
      aria-hidden="true"
      class="_icon_729a27"
    >
      â–¼
    </span>
  </button>
</div>
`;

exports[`LanguageSelector renders correctly with custom languages 1`] = `
<div
  class="_selector_729a27"
>
  <button
    class="_trigger_729a27"
    type="button"
  >
    <span
      class="_value_729a27"
    >
      Select a language
    </span>
    <span
      aria-hidden="true"
      class="_icon_729a27"
    >
      â–¼
    </span>
  </button>
</div>
`;

exports[`LanguageSelector renders correctly with default props 1`] = `
<div
  class="_selector_729a27"
>
  <button
    class="_trigger_729a27"
    type="button"
  >
    <span
      class="_value_729a27"
    >
      Select a language
    </span>
    <span
      aria-hidden="true"
      class="_icon_729a27"
    >
      â–¼
    </span>
  </button>
</div>
`;

exports[`LanguageSelector renders correctly with selected language 1`] = `
<div
  class="_selector_729a27"
>
  <button
    class="_trigger_729a27"
    type="button"
  >
    <span
      class="_value_729a27"
    >
      ðŸ‡ºðŸ‡¸ English
    </span>
    <span
      aria-hidden="true"
      class="_icon_729a27"
    >
      â–¼
    </span>
  </button>
</div>
`;

exports[`LanguageSelector renders correctly without flags 1`] = `
<div
  class="_selector_729a27"
>
  <button
    class="_trigger_729a27"
    type="button"
  >
    <span
      class="_value_729a27"
    >
      Select a language
    </span>
    <span
      aria-hidden="true"
      class="_icon_729a27"
    >
      â–¼
    </span>
  </button>
</div>
`;
